src/dbnode/storage/namespace/schema.go

```diff
@@ -0,0 +1,226 @@
+	options  *nsproto.SchemaOptions
+	latestId string
+	// a map of schema version to schema descriptor.
+	versions map[string]*schemaDescr
+}
+
+func (sr *schemaRegistry) Equal(o SchemaRegistry) bool {
+	var osr *schemaRegistry
+	var ok bool
+	if sr == nil && o == nil {
+		return true
+	}
+	if sr != nil && o == nil || sr == nil && o != nil {
+		return false
+	}
+
+	if osr, ok = o.(*schemaRegistry); !ok {
+		return false
+	}
+	// compare latest version
+	if sr.latestId != osr.latestId {
+		return false
+	}
+
+	// compare version map
+	if len(sr.versions) != len(osr.versions) {
+		return false
+	}
+	for v, sd := range sr.versions {
+		osd, ok := osr.versions[v]
+		if !ok {
+			return false
+		}
+		if !sd.Equal(osd) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (sr *schemaRegistry) Get(id string) (SchemaDescr, error) {
+	sd, ok := sr.versions[id]
+	if !ok {
+		return nil, errSchemaNotFound
+	}
+	return sd, nil
+}
+
+func (sr *schemaRegistry) GetLatest() (SchemaDescr, error) {
+	return sr.Get(sr.latestId)
+}
+
+// toSchemaOptions returns the corresponding SchemaOptions proto for the provided SchemaRegistry
+func toSchemaOptions(sr SchemaRegistry) *nsproto.SchemaOptions {
+	if sr == nil {
+		return nil
+	}
+	_, ok := sr.(*schemaRegistry)
+	if !ok {
+		return nil
+	}
+	return sr.(*schemaRegistry).options
+}
+
+func emptySchemaRegistry() SchemaRegistry {
+	return &schemaRegistry{options: nil, versions: make(map[string]*schemaDescr)}
+}
+
+// LoadSchemaRegistry loads schema registry from SchemaOptions proto.
+func LoadSchemaRegistry(options *nsproto.SchemaOptions) (SchemaRegistry, error) {
+	sr := &schemaRegistry{options: options, versions: make(map[string]*schemaDescr)}
+	if options == nil ||
+		options.GetHistory() == nil ||
+		len(options.GetHistory().GetVersions()) == 0 {
+		return sr, nil
+	}
+
+	msgName := options.GetDefaultMessageName()
+	if len(msgName) == 0 {
+		return nil, xerrors.Wrap(errInvalidSchemaOptions, "default message name is not specified")
+	}
+
+	var prevId string
+	for _, fdbSet := range options.GetHistory().GetVersions() {
+		if len(prevId) > 0 && fdbSet.PrevId != prevId {
+			return nil, xerrors.Wrapf(errInvalidSchemaOptions, "schema history is not sorted by deploy id in ascending order")
+		}
+		sd, err := loadFileDescriptorSet(fdbSet, msgName)
+		if err != nil {
+			return nil, err
+		}
+		sr.versions[sd.DeployId()] = sd
+		prevId = sd.DeployId()
+	}
+	sr.latestId = prevId
+
+	return sr, nil
+}
+
+func loadFileDescriptorSet(fdSet *nsproto.FileDescriptorSet, msgName string) (*schemaDescr, error) {
+	// assuming file descriptors are topological sorted
+	var dependencies []*desc.FileDescriptor
+	var curfd *desc.FileDescriptor
+	for i, fdb := range fdSet.Descriptors {
+		fdp, err := decodeProtoFileDescriptor(fdb)
+		if err != nil {
+			return nil, xerrors.Wrapf(err, "failed to decode file descriptor(%d) in version(%d)", i, fdSet.DeployId)
+		}
+		fd, err := desc.CreateFileDescriptor(fdp, dependencies...)
```

Comment:
Ah I see....where will we make sure things are in the right order? If things aren't sorted properly will this return an error?

Reply:
yes, this will, I will add a negative test